README.md
# fluent-plugin-gcloud-pubsub-custom

![Test](https://github.com/mia-0032/fluent-plugin-gcloud-pubsub-custom/workflows/Test/badge.svg)
[![Gem Version](https://badge.fury.io/rb/fluent-plugin-gcloud-pubsub-custom.svg)](http://badge.fury.io/rb/fluent-plugin-gcloud-pubsub-custom)

This plugin is forked from https://github.com/mdoi/fluent-plugin-gcloud-pubsub

## Overview

[Google Cloud Pub/Sub](https://cloud.google.com/pubsub/) Input/Output(BufferedOutput) plugin for [Fluentd](http://www.fluentd.org/) with [google-cloud](https://googlecloudplatform.github.io/google-cloud-ruby/) gem

- Publish messages to Google Cloud Pub/Sub
- Pull messages from Google Cloud Pub/Sub

## Preparation

- Create a project on Google Developer Console
- Add a topic of Cloud Pub/Sub to the project
- Add a pull style subscription to the topic
- Download your credential (json) or [set scope on GCE instance](https://cloud.google.com/compute/docs/api/how-tos/authorization)

When using output plugin, you need to grant Pub/Sub Publisher and Pub/Sub Viewer role to IAM.

Also, when using input plugin, you need to grant Pub/Sub Subscriber and Pub/Sub Viewer role to IAM.

## Requirements

| fluent-plugin-gcloud-pubsub-custom | fluentd | ruby |
|------------------------|---------|------|
| >= 1.0.0 | >= v0.14.0 | >= 2.1 |
|  < 1.0.0 | >= v0.12.0 | >= 1.9 |

## Installation

Install by gem:

```shell
$ gem install fluent-plugin-gcloud-pubsub-custom
```

**Caution**

This plugin doesn't work in [td-agent](http://docs.fluentd.org/articles/install-by-rpm).

Please use in [Fluentd installed by gem](http://docs.fluentd.org/articles/install-by-gem).

## Configuration

### Publish messages

Use `gcloud_pubsub` output plugin.

```
<match example.publish>
  @type gcloud_pubsub
  project <YOUR PROJECT>
  key <YOUR KEY>
  topic <YOUR TOPIC>
  autocreate_topic false
  max_messages 1000
  max_total_size 9800000
  max_message_size 4000000
  endpoint <Endpoint URL>
  timeout 60
  compress_batches false
  <buffer>
    @type memory
    flush_interval 1s
  </buffer>
  <format>
    @type json
  </format>
</match>
```

- `project` (optional)
  - Set your GCP project.
  - Running fluentd on GCP, you don't have to specify.
  - You can also use environment variable such as `GCLOUD_PROJECT`.
- `key` (optional)
  - Set your credential file path.
  - Running fluentd on GCP, you can use scope instead of specifying this.
  - You can also use environment variable such as `GCLOUD_KEYFILE`.
- `topic` (required)
  - Set topic name to publish.
  - You can use placeholder in this param. See: https://docs.fluentd.org/v1.0/articles/buffer-section
- `dest_project` (optional, default: `nil`)
  - Set your destination GCP project if you publish messages cross project.
- `autocreate_topic` (optional, default: `false`)
  - If set to `true`, specified topic will be created when it doesn't exist.
- `max_messages` (optional, default: `1000`)
  - Publishing messages count per request to Cloud Pub/Sub.
    - See https://cloud.google.com/pubsub/quotas#other_limits
- `max_total_size` (optional, default: `9800000` = `9.8MB`)
  - Publishing messages bytesize per request to Cloud Pub/Sub. This parameter affects only message size. You should specify a little smaller value than quota.
    - See https://cloud.google.com/pubsub/quotas#other_limits
- `max_message_size` (optional, default: `4000000` = `4MB`)
  - Messages exceeding `max_message_size` are not published because Pub/Sub clients cannot receive it.
- `attribute_keys` (optional, default: `[]`)
  - Extract these fields from the record and send them as attributes on the Pub/Sub message. Cannot be set if compress_batches is enabled.
- `attribute_key_values` (optional, default: `{}`)
  - Publishing the set fields as attributes generated from input params
- `endpoint`(optional)
  - Set Pub/Sub service endpoint. For more information, see [Service Endpoints](https://cloud.google.com/pubsub/docs/reference/service_apis_overview#service_endpoints)
- `compression` (optional, default: `nil`)
  - If set to `gzip`, messages will be compressed with gzip.
- `timeout` (optional)
  - Set default timeout to use in publish requests.
  - Publishing the set fields as attributes.
- `metric_prefix` (optional, default: `fluentd_output_gcloud_pubsub`)
  - The prefix for Prometheus metric names
- `compress_batches` (optional, default: `false`)
  - If set to `true`, messages will be batched and compressed before publication. See [message compression](#message-compression) for details.

### Pull messages

Use `gcloud_pubsub` input plugin.

```
<source>
  @type gcloud_pubsub
  tag example.pull
  project <YOUR PROJECT>
  key <YOUR KEY>
  topic <YOUR TOPIC>
  subscription <YOUR SUBSCRIPTION>
  max_messages 1000
  return_immediately true
  pull_interval 0.5
  pull_threads 2
  parse_error_action exception
  enable_rpc true
  rpc_bind 0.0.0.0
  rpc_port 24680
  <parse>
    @type json
  </parse>
</source>
```

- `tag` (required)
  - Set tag of messages.
  - If `tag_key` is specified, `tag` is used as tag when record don't have specified key.
- `tag_key` (optional)
  - Set key to be used as tag.
- `project` (optional)
  - Set your GCP project
  - Running fluentd on GCP, you don't have to specify.
  - You can also use environment variable such as `GCLOUD_PROJECT`.
- `key` (optional)
  - Set your credential file path.
  - Running fluentd on GCP, you can use scope instead of specifying this.
  - You can also use environment variable such as `GCLOUD_KEYFILE`.
- `topic` (optional)
  - Set topic name that the subscription belongs to.
- `subscription` (required)
  - Set subscription name to pull.
- `max_messages` (optional, default: `100`)
  - See maxMessages on https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/pull
- `return_immediately` (optional, default: `true`)
  - See returnImmediately on https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/pull
  - If `return_immediately` is `true` or pulling message is stopped by HTTP RPC, this plugin wait `pull_interval` each pull.
- `pull_interval` (optional, default: `5.0`)
  - Pulling messages by intervals of specified seconds.
- `pull_threads` (optional, default: `1`)
  - Set number of threads to pull messages.
- `attribute_keys` (optional, default: `[]`)
  - Acquire these fields from attributes on the Pub/Sub message and merge them into the record.
- `parse_error_action` (optional, default: `exception`)
  - Set error type when parsing messages fails.
    - `exception`: Raise exception. Messages are not acknowledged.
    - `warning`: Only logging as warning.
- `metric_prefix` (optional, default: `fluentd_input_gcloud_pubsub`)
  - The prefix for Prometheus metric names
- `enable_rpc` (optional, default: `false`)
  - If `true` is specified, HTTP RPC to stop or start pulling message is enabled.
- `rpc_bind` (optional, default: `0.0.0.0`)
  - Bind IP address for HTTP RPC.
- `rpc_port` (optional, default: `24680`)
  - Port for HTTP RPC.
- `decompression` (optional, default: `nil`)
  - If set to `gzip`, messages will be decompressed with gzip.

## Message compression

The `compress_batches` option can be used to enable the compression of messages
_before_ publication to Pub/Sub.

This works by collecting the buffered messages, taking up to `max_total_size` or
`max_message_size` input records, then compressing them with Zlib (i.e.
gzip/Deflate) before publishing them as a single message to the Pub/Sub topic.

When transporting large volumes of records via Pub/Sub, e.g. multiple Terabytes
per month, this can lead to significant cost savings, as typically the CPU time
required to compress the messages will be minimal in comparison to the Pub/Sub
costs.

The compression ratio achievable will vary largely depending on the homogeneity
of the input records, but typically will be 50% at the very minimum and often
around 80-90%.

In order to achieve good compression, consider the following:
- Ensure that the buffer is being filled with a reasonable batch of messages: do
  not use `flush_mode immediate`, and keep the `flush_interval` value
  sufficiently high. Use the Prometheus metrics to determine how many records
  are being published per message.
- Keep the `max_messages` and `max_message_size` values high (the defaults are
  optimal).
- If there are many different sources of messages being mixed and routed to a
  single `gcloud_pubsub` output, use multiple outputs (which will each have
  their own buffer) through tagging or [labelling][fluentd-labels].

[fluentd-labels]: https://docs.fluentd.org/quickstart/life-of-a-fluentd-event#labels

The receiving end must be able to decode these compressed batches of messages,
which it can determine via an attribute set on the Pub/Sub message. The
`gcloud_pubsub` input plugin will do this transparently, decompressing any
messages which contain a batch of records and normally processing any messages
which represent just a single record.
Therefore, as long as all of the receivers are updated with support for
compressed batches first, it's then possible to gradually roll out this feature.

## Prometheus metrics

The input and output plugins expose several metrics in order to monitor
performance:

- `fluentd_output_gcloud_pubsub_compression_enabled`
  - Gauge: Whether compression/batching is enabled
- `fluentd_output_gcloud_pubsub_messages_published_per_batch`
  - Histogram: Number of records published to Pub/Sub per buffer flush
- `fluentd_output_gcloud_pubsub_messages_published_bytes`
  - Histogram: Total size in bytes of the records published to Pub/Sub,
    **before** compression.
- `fluentd_output_gcloud_pubsub_messages_compression_duration_seconds`
  - Histogram: Time taken to compress a batch of messages
- `fluentd_output_gcloud_pubsub_messages_compressed_size_per_original_size_ratio`
  - Histogram: Compression ratio achieved on a batch of messages, expressed in
    terms of space saved.

- `fluentd_input_gcloud_pubsub_pull_errors_total`
  - Counter: Errors encountered while pulling or processing messages (split by a
    `retryable` label)
- `fluentd_input_gcloud_pubsub_messages_pulled`
  - Histogram: Number of Pub/Sub messages pulled by the subscriber on each invocation
- `fluentd_input_gcloud_pubsub_messages_pulled_bytes`
  - Histogram: Total size in bytes of the Pub/Sub messages pulled by the subscriber on each invocation

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

## TODO

- Add `tag` attribute in output plugin and use `tag` attribute as tag in input plugin.
- Send ack after other output plugin committed (if possible).

## Authors

- [@mdoi](https://github.com/mdoi) : First author
- [@mia-0032](https://github.com/mia-0032) : This version author


CHANGELOG.md
## ChangeLog

### Release 1.3.3 - 2021/01/05

- Add support for payload compression with `compress_batches`

### Release 1.3.2 - 2019/08/16

- Input plugin
  - Make `topic` parameter optional

### Release 1.3.1 - 2018/06/08

- Input plugin
  - Dumps raw string to logs for parsing errors

### Release 1.3.0 - 2018/05/18

- Make attributes available 

### Release 1.2.0 - 2018/04/24

- Output plugin
  - Support inject section 

### Release 1.1.0 - 2018/03/30

- Output plugin
  - The placeholder is now available in topic param

### Release 1.0.3 - 2018/03/29

- Bump up google-cloud-pubsub to v0.30.x
- Make config file Fluentd v1 compatible

### Release 1.0.2 - 2017/09/11

- Bump up google-cloud-pubsub to v0.27.x

### Release 1.0.1 - 2017/09/05

- Bump up google-cloud-pubsub to v0.26.x

### Release 1.0.0 - 2017/06/23

- Fluentd v0.14 ready
  - Fluentd v0.12 is not supported in the later version

### Release 0.4.6 - 2017/05/14

- Output plugin
  - Make messages exceeding configured size not be published because Pub/Sub clients cannot receive it

### Release 0.4.5 - 2017/04/02

- Bump up google-cloud-pubsub to v0.24.x

### Release 0.4.4 - 2017/03/07

- Bump up google-cloud-pubsub to v0.23.x

### Release 0.4.3 - 2017/02/16

- Input plugin
  - Add "status" method to the http rpc api

### Release 0.4.2 - 2017/02/03

- Make retry to get topic/subscription when Pub/Sub API returns 50x code

### Release 0.4.1 - 2017/02/02

- Bump up google-cloud-pubsub to v0.22.x
- Input plugin
  - Rescue 50x errors on acknowledge api

### Release 0.4.0 - 2017/01/21

- Input plugin
  - Add feature to use record key as tag

### Release 0.3.4 - 2017/01/03

- Output plugin
  - Rescue 50x errors
- Input plugin
  - Guard emit to be called with multi-threading
  - Rescue 50x errors
  - Enabled to select whether to raise an exception if message processing failed

### Release 0.3.3 - 2016/12/03

- Input plugin
  - Fix undefined variable error

### Release 0.3.2 - 2016/11/13

- Add plugin param desc
- Input plugin
  - Improve handling to acknowledge messages

### Release 0.3.1 - 2016/11/03

- Output plugin
  - Improve error handling

### Release 0.3.0 - 2016/10/30

- Bump up google-cloud-pubsub to v0.21
- Input plugin
  - Add multithreaded pulling feature

### Release 0.2.0 - 2016/10/15

- Input plugin
  - Add HTTP RPC feature

### Release 0.1.4 - 2016/09/19

- Input plugin
  - `pull_interval` can be specified float value
  - `topic` must be specified

### Release 0.1.3 - 2016/09/17

- Input plugin
  - Fix error handling and add debug logging

### Release 0.1.2 - 2016/09/11

- Output plugin
  - Change default max message size and add debug message

### Release 0.1.1 - 2016/08/27

- Bump up google-cloud-pubsub (gcloud-ruby) to 0.20

### Release 0.1.0 - 2016/08/22

- Use formatter / parser plugin and add format configuration
- Bump up gcloud-ruby to 0.12
- Remove dependency on lightening buffer
- Fix error caused by Pub/Sub quotas


test/helper.rb
# frozen_string_literal: true

require "rubygems"
require "bundler"
begin
  Bundler.setup(:default, :development)
rescue Bundler::BundlerError => e
  warn e.message
  warn "Run `bundle install` to install missing gems"
  exit e.status_code
end
require "test/unit"
require "test/unit/rr"

$LOAD_PATH.unshift(File.expand_path("../../", __FILE__))

require "fluent/test"
require "fluent/test/driver/output"
require "fluent/test/helpers"

require "fluent/plugin/in_gcloud_pubsub"
require "fluent/plugin/out_gcloud_pubsub"

unless ENV.key?("VERBOSE")
  nulllogger = Object.new
  nulllogger.instance_eval do |_obj|
    def method_missing(method, *args)
      # pass
    end
  end
  $log = nulllogger
end

Test::Unit::TestCase.include(Fluent::Test::Helpers)
Test::Unit::TestCase.extend(Fluent::Test::Helpers)


test/test_helper.rb
# frozen_string_literal: true

require "rubygems"
require "bundler"
begin
  Bundler.setup(:default, :development)
rescue Bundler::BundlerError => e
  warn e.message
  warn "Run `bundle install` to install missing gems"
  exit e.status_code
end
require "test/unit"
require "test/unit/rr"

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), "..", "lib"))
$LOAD_PATH.unshift(File.dirname(__FILE__))

require "fluent/test"
unless ENV.key?("VERBOSE")
  nulllogger = Object.new
  nulllogger.instance_eval do |_obj|
    def method_missing(method, *args)
      # pass
    end
  end
  $log = nulllogger
end

require "fluent/plugin/in_gcloud_pubsub"
require "fluent/plugin/out_gcloud_pubsub"

class Test::Unit::TestCase
end


lib/fluent/plugin/out_gcloud_pubsub.rb
# frozen_string_literal: true

require "fluent/plugin/output"
require "fluent/plugin/gcloud_pubsub/client"
require "fluent/plugin/gcloud_pubsub/metrics"
require "fluent/plugin_helper/inject"
require "prometheus/client"

module Fluent::Plugin
  class GcloudPubSubOutput < Output
    include Fluent::PluginHelper::Inject

    Fluent::Plugin.register_output("gcloud_pubsub", self)

    helpers :compat_parameters, :formatter

    DEFAULT_BUFFER_TYPE = "memory"
    DEFAULT_FORMATTER_TYPE = "json"

    desc "Set your GCP project."
    config_param :project, :string, default: nil
    desc "Set your credential file path."
    config_param :key, :string, default: nil
    desc "Set topic name to publish."
    config_param :topic, :string
    desc "If set to `true`, specified topic will be created when it doesn't exist."
    config_param :autocreate_topic, :bool, default: false
    desc "Publishing messages count per request to Cloud Pub/Sub."
    config_param :max_messages, :integer, default: 1000
    desc "Publishing messages bytesize per request to Cloud Pub/Sub."
    config_param :max_total_size, :integer, default: 9_800_000 # 9.8MB
    desc "Limit bytesize per message."
    config_param :max_message_size, :integer, default: 4_000_000 # 4MB
    desc "Extract these fields from the record and send them as attributes on the Pub/Sub message. " \
         "Cannot be set if compress_batches is enabled."
    config_param :attribute_keys, :array, default: []
    desc "The prefix for Prometheus metric names"
    config_param :metric_prefix, :string, default: "fluentd_output_gcloud_pubsub"
    desc "If set to `true`, messages will be batched and compressed before publication"
    config_param :compress_batches, :bool, default: false

    config_section :buffer do
      config_set_default :@type, DEFAULT_BUFFER_TYPE
    end

    config_section :format do
      config_set_default :@type, DEFAULT_FORMATTER_TYPE
    end

    # rubocop:disable Metrics/MethodLength
    def configure(conf)
      compat_parameters_convert(conf, :buffer, :formatter)
      super
      placeholder_validate!(:topic, @topic)
      @formatter = formatter_create

      if @compress_batches && !@attribute_keys.empty?
        # The attribute_keys option is implemented by extracting keys from the
        # record and setting them on the Pub/Sub message.
        # This is not possible in compressed mode, because we're sending just a
        # single Pub/Sub message that comprises many records, therefore the
        # attribute keys would clash.
        raise Fluent::ConfigError, ":attribute_keys cannot be used when compression is enabled"
      end

      @messages_published =
        Fluent::GcloudPubSub::Metrics.register_or_existing(:"#{@metric_prefix}_messages_published_per_batch") do
          ::Prometheus::Client.registry.histogram(
            :"#{@metric_prefix}_messages_published_per_batch",
            "Number of records published to Pub/Sub per buffer flush",
            {},
            [1, 10, 50, 100, 250, 500, 1000],
          )
        end

      @bytes_published =
        Fluent::GcloudPubSub::Metrics.register_or_existing(:"#{@metric_prefix}_messages_published_bytes") do
          ::Prometheus::Client.registry.histogram(
            :"#{@metric_prefix}_messages_published_bytes",
            "Total size in bytes of the records published to Pub/Sub",
            {},
            [100, 1000, 10_000, 100_000, 1_000_000, 5_000_000, 10_000_000],
          )
        end

      @compression_enabled =
        Fluent::GcloudPubSub::Metrics.register_or_existing(:"#{@metric_prefix}_compression_enabled") do
          ::Prometheus::Client.registry.gauge(
            :"#{@metric_prefix}_compression_enabled",
            "Whether compression/batching is enabled",
            {},
          )
        end
      @compression_enabled.set(common_labels, @compress_batches ? 1 : 0)
    end
    # rubocop:enable Metrics/MethodLength

    def start
      super
      @publisher = Fluent::GcloudPubSub::Publisher.new @project, @key, @autocreate_topic, @metric_prefix
    end

    def format(tag, time, record)
      record = inject_values_to_record(tag, time, record)
      attributes = {}
      @attribute_keys.each do |key|
        attributes[key] = record.delete(key)
      end
      [@formatter.format(tag, time, record), attributes].to_msgpack
    end

    def formatted_to_msgpack_binary?
      true
    end

    def multi_workers_ready?
      true
    end

    def write(chunk)
      topic = extract_placeholders(@topic, chunk.metadata)

      messages = []
      size = 0

      chunk.msgpack_each do |msg, attr|
        msg = Fluent::GcloudPubSub::Message.new(msg, attr)
        if msg.bytesize > @max_message_size
          log.warn "Drop a message because its size exceeds `max_message_size`", size: msg.bytesize
          next
        end
        if messages.length + 1 > @max_messages || size + msg.bytesize > @max_total_size
          publish(topic, messages)
          messages = []
          size = 0
        end
        messages << msg
        size += msg.bytesize
      end

      publish(topic, messages) unless messages.empty?
    rescue Fluent::GcloudPubSub::RetryableError => e
      log.warn "Retryable error occurs. Fluentd will retry.", error_message: e.to_s, error_class: e.class.to_s
      raise e
    rescue StandardError => e
      log.error "unexpected error", error_message: e.to_s, error_class: e.class.to_s
      log.error_backtrace
      raise e
    end

    private

    def publish(topic, messages)
      size = messages.map(&:bytesize).inject(:+)
      log.debug "send message topic:#{topic} length:#{messages.length} size:#{size}"

      @messages_published.observe(common_labels, messages.length)
      @bytes_published.observe(common_labels, size)

      @publisher.publish(topic, messages, @compress_batches)
    end

    def common_labels
      { topic: @topic }
    end
  end
end


lib/fluent/plugin/in_gcloud_pubsub.rb
# frozen_string_literal: true

require "json"
require "webrick"

require "fluent/plugin/input"
require "fluent/plugin/parser"

require "fluent/plugin/gcloud_pubsub/client"
require "fluent/plugin/gcloud_pubsub/metrics"

module Fluent::Plugin
  class GcloudPubSubInput < Input
    Fluent::Plugin.register_input("gcloud_pubsub", self)

    helpers :compat_parameters, :parser, :thread

    DEFAULT_PARSER_TYPE = "json"

    class FailedParseError < StandardError
    end

    desc "Set tag of messages."
    config_param :tag, :string
    desc "Set key to be used as tag."
    config_param :tag_key, :string, default: nil
    desc "Set your GCP project."
    config_param :project, :string, default: nil
    desc "Set your credential file path."
    config_param :key, :string, default: nil
    desc "Set topic name to pull."
    config_param :topic, :string, default: nil
    desc "Set subscription name to pull."
    config_param :subscription, :string
    desc "Pulling messages by intervals of specified seconds."
    config_param :pull_interval, :float, default: 5.0
    desc "Max messages pulling at once."
    config_param :max_messages, :integer, default: 100
    desc "Setting `true`, keepalive connection to wait for new messages."
    config_param :return_immediately, :bool, default: true
    desc "Set number of threads to pull messages."
    config_param :pull_threads, :integer, default: 1
    desc "Acquire these fields from attributes on the Pub/Sub message and merge them into the record"
    config_param :attribute_keys,     :array, default: []
    desc "Set error type when parsing messages fails."
    config_param :parse_error_action, :enum, default: :exception, list: %i[exception warning]
    desc "The prefix for Prometheus metric names"
    config_param :metric_prefix, :string, default: "fluentd_input_gcloud_pubsub"
    # for HTTP RPC
    desc "If `true` is specified, HTTP RPC to stop or start pulling message is enabled."
    config_param :enable_rpc, :bool, default: false
    desc "Bind IP address for HTTP RPC."
    config_param :rpc_bind, :string, default: "0.0.0.0"
    desc "Port for HTTP RPC."
    config_param :rpc_port, :integer, default: 24_680

    config_section :parse do
      config_set_default :@type, DEFAULT_PARSER_TYPE
    end

    class RPCServlet < WEBrick::HTTPServlet::AbstractServlet
      class Error < StandardError; end

      def initialize(server, plugin)
        super
        @plugin = plugin
      end

      # rubocop:disable Naming/MethodName
      def do_GET(req, res)
        begin
          code, header, body = process(req, res)
        rescue StandardError
          code, header, body = render_json(500, {
                                             "ok" => false,
                                             "message" => "Internal Server Error",
                                             "error" => $ERROR_INFO.to_s,
                                             "backtrace" => $ERROR_INFO.backtrace,
                                           })
        end

        res.status = code
        header.each_pair do |k, v|
          res[k] = v
        end
        res.body = body
      end
      # rubocop:enable Naming/MethodName

      def render_json(code, obj)
        [code, { "Content-Type" => "application/json" }, obj.to_json]
      end

      def process(req, _res)
        ret = { "ok" => true }
        case req.path_info
        when "/stop"
          @plugin.stop_pull
        when "/start"
          @plugin.start_pull
        when "/status"
          ret["status"] = @plugin.status_of_pull
        else
          raise Error, "Invalid path_info: #{req.path_info}"
        end
        render_json(200, ret)
      end
    end

    # rubocop:disable Metrics/MethodLength
    def configure(conf)
      compat_parameters_convert(conf, :parser)
      super
      @rpc_srv = nil
      @rpc_thread = nil
      @stop_pull = false

      @extract_tag = if @tag_key.nil?
                       method(:static_tag)
                     else
                       method(:dynamic_tag)
                     end

      @parser = parser_create

      @messages_pulled =
        Fluent::GcloudPubSub::Metrics.register_or_existing(:"#{@metric_prefix}_messages_pulled") do
          ::Prometheus::Client.registry.histogram(
            :"#{@metric_prefix}_messages_pulled",
            "Number of Pub/Sub messages pulled by the subscriber on each invocation",
            {},
            [0, 1, 10, 50, 100, 250, 500, 1000],
          )
        end

      @messages_pulled_bytes =
        Fluent::GcloudPubSub::Metrics.register_or_existing(:"#{@metric_prefix}_messages_pulled_bytes") do
          ::Prometheus::Client.registry.histogram(
            :"#{@metric_prefix}_messages_pulled_bytes",
            "Total size in bytes of the Pub/Sub messages pulled by the subscriber on each invocation",
            {},
            [100, 1000, 10_000, 100_000, 1_000_000, 5_000_000, 10_000_000],
          )
        end

      @pull_errors =
        Fluent::GcloudPubSub::Metrics.register_or_existing(:"#{@metric_prefix}_pull_errors_total") do
          ::Prometheus::Client.registry.counter(
            :"#{@metric_prefix}_pull_errors_total",
            "Errors encountered while pulling or processing messages",
            {},
          )
        end
    end
    # rubocop:enable Metrics/MethodLength

    def start
      super
      start_rpc if @enable_rpc

      @subscriber = Fluent::GcloudPubSub::Subscriber.new @project, @key, @topic, @subscription
      log.debug "connected subscription:#{@subscription} in project #{@project}"

      @emit_guard = Mutex.new
      @stop_subscribing = false
      @subscribe_threads = []
      @pull_threads.times do |idx|
        @subscribe_threads.push thread_create("in_gcloud_pubsub_subscribe_#{idx}".to_sym, &method(:subscribe))
      end
    end

    def shutdown
      if @rpc_srv
        @rpc_srv.shutdown
        @rpc_srv = nil
      end
      @rpc_thread = nil if @rpc_thread
      @stop_subscribing = true
      @subscribe_threads.each(&:join)
      super
    end

    def stop_pull
      @stop_pull = true
      log.info "stop pull from subscription:#{@subscription}"
    end

    def start_pull
      @stop_pull = false
      log.info "start pull from subscription:#{@subscription}"
    end

    def status_of_pull
      @stop_pull ? "stopped" : "started"
    end

    private

    def static_tag(_record)
      @tag
    end

    def dynamic_tag(record)
      record.delete(@tag_key) || @tag
    end

    def start_rpc
      log.info "listening http rpc server on http://#{@rpc_bind}:#{@rpc_port}/"
      @rpc_srv = WEBrick::HTTPServer.new(
        {
          BindAddress: @rpc_bind,
          Port: @rpc_port,
          Logger: WEBrick::Log.new(STDERR, WEBrick::Log::FATAL),
          AccessLog: [],
        },
      )
      @rpc_srv.mount("/api/in_gcloud_pubsub/pull/", RPCServlet, self)
      @rpc_thread = thread_create(:in_gcloud_pubsub_rpc_thread) do
        @rpc_srv.start
      end
    end

    def subscribe
      until @stop_subscribing
        _subscribe unless @stop_pull

        sleep @pull_interval if @return_immediately || @stop_pull
      end
    rescue StandardError => e
      log.error "unexpected error", error_message: e.to_s, error_class: e.class.to_s
      log.error_backtrace e.backtrace
    end

    def _subscribe
      messages = @subscriber.pull @return_immediately, @max_messages
      @messages_pulled.observe(common_labels, messages.size)
      if messages.empty?
        log.debug "no messages are pulled"
        return
      end

      messages_size = messages.sum do |message|
        message.data.bytesize + message.attributes.sum { |k, v| k.bytesize + v.bytesize }
      end
      @messages_pulled_bytes.observe(common_labels, messages_size)

      process messages
      @subscriber.acknowledge messages

      log.debug "#{messages.length} message(s) processed"
    rescue Fluent::GcloudPubSub::RetryableError => e
      @pull_errors.increment(common_labels.merge({ retryable: true }))
      log.warn "Retryable error occurs. Fluentd will retry.", error_message: e.to_s, error_class: e.class.to_s
    rescue StandardError => e
      @pull_errors.increment(common_labels.merge({ retryable: false }))
      log.error "unexpected error", error_message: e.to_s, error_class: e.class.to_s
      log.error_backtrace e.backtrace
    end

    def process(messages)
      event_streams = Hash.new do |hsh, key|
        hsh[key] = Fluent::MultiEventStream.new
      end

      messages.each do |m|
        lines_attributes = Fluent::GcloudPubSub::MessageUnpacker.unpack(m)

        lines_attributes.each do |line, attributes|
          @parser.parse(line) do |time, record|
            if time && record
              @attribute_keys.each do |key|
                record[key] = attributes[key]
              end

              event_streams[@extract_tag.call(record)].add(time, record)
            else
              case @parse_error_action
              when :exception
                raise FailedParseError, "pattern not match: #{line}"
              else
                log.warn "pattern not match", record: line
              end
            end
          end
        end
      end

      event_streams.each do |tag, es|
        # There are some output plugins not to supposed to be called with multi-threading.
        # Maybe remove in the future.
        @emit_guard.synchronize do
          router.emit_stream(tag, es)
        end
      end
    end

    def common_labels
      { subscription: @subscription }
    end
  end
end
